File: packages/web/src/api/manifest.ts
import type { PackSummary } from "./packs";

export interface PermittedException {
  term: string;
  reason: string;
}

export interface Banished {
  words: string[];
  phrases: string[];
  permitted_exceptions: PermittedException[];
}

export interface Rules {
  no_em_dashes: boolean;
  no_ascii_double_hyphen_between_letters: boolean;
  no_sentence_starters: string[];
}

export interface PopCulture {
  allowed: string[];
  banned: string[];
}

export interface FormatEntry {
  name: string;
  file: string;
  description?: string | null;
}

export interface SampleEntry {
  id: string;
  file: string;
  description?: string | null;
}

export interface BioEntry {
  name: string;
  file: string;
  max_chars?: number | null;
  target_words?: number | null;
  third_person?: boolean;
  description?: string | null;
}

export interface PackInfo {
  slug: string;
  name: string;
  version: string;
  author: string;
  description?: string | null;
  homepage?: string | null;
}

export interface Persona {
  identity: string;
  one_line: string;
}

export interface Manifest {
  spec_version: "1.0";
  pack: PackInfo;
  persona: Persona;
  banished: Banished;
  rules: Rules;
  pop_culture: PopCulture;
  formats: FormatEntry[];
  samples: SampleEntry[];
  bios: BioEntry[];
}

export class ManifestValidationError extends Error {
  constructor(
    public readonly fieldErrors: Record<string, string>,
    message: string,
  ) {
    super(message);
    this.name = "ManifestValidationError";
  }
}

export async function putManifest(slug: string, manifest: Manifest): Promise<PackSummary> {
  const response = await fetch(`/api/packs/${encodeURIComponent(slug)}/manifest`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(manifest),
  });

  if (response.status === 422) {
    const body = (await response.json()) as {
      detail?: { errors?: Array<{ path: string; message: string }> };
    };
    const errList = body?.detail?.errors;
    if (errList) {
      const fieldErrors: Record<string, string> = {};
      for (const e of errList) fieldErrors[e.path] = e.message;
      throw new ManifestValidationError(
        fieldErrors,
        `Validation failed: ${errList.length} error(s).`,
      );
    }
    throw new Error(`HTTP 422 on /api/packs/${slug}/manifest`);
  }

  if (!response.ok) {
    throw new Error(`HTTP ${response.status} on /api/packs/${slug}/manifest`);
  }

  return (await response.json()) as PackSummary;
}

File: packages/api/myvoice/packs/manifest.py
"""Pydantic models for a Style Pack manifest (`stylepack.yaml`).

Models match SPEC.md v1.0. Unknown fields are rejected (model_config
forbids extra). This catches typos like `no_em_dash` vs `no_em_dashes`.
"""

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class _StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class Pack(_StrictModel):
    slug: str = Field(min_length=1)
    name: str = Field(min_length=1)
    version: str = Field(min_length=1)
    author: str = Field(min_length=1)
    description: str | None = None
    homepage: str | None = None


class Persona(_StrictModel):
    identity: str = Field(min_length=1)
    one_line: str = Field(min_length=1)
    tone: str | None = None


class PermittedException(_StrictModel):
    term: str = Field(min_length=1)
    reason: str = Field(min_length=1)


class Banished(_StrictModel):
    words: list[str] = Field(default_factory=list)
    phrases: list[str] = Field(default_factory=list)
    permitted_exceptions: list[PermittedException] = Field(default_factory=list)


class Rules(_StrictModel):
    no_em_dashes: bool = True
    no_ascii_double_hyphen_between_letters: bool = True
    no_sentence_starters: list[str] = Field(default_factory=list)


class PopCulture(_StrictModel):
    allowed: list[str] = Field(default_factory=list)
    banned: list[str] = Field(default_factory=list)


class Format(_StrictModel):
    name: str = Field(min_length=1)
    file: str = Field(min_length=1)
    description: str | None = None


class Sample(_StrictModel):
    id: str = Field(min_length=1)
    file: str = Field(min_length=1)
    description: str | None = None


class Bio(_StrictModel):
    name: str = Field(min_length=1)
    file: str = Field(min_length=1)
    max_chars: int | None = Field(default=None, gt=0)
    target_words: int | None = Field(default=None, gt=0)
    third_person: bool = False
    description: str | None = None


class Manifest(_StrictModel):
    spec_version: Literal["1.0"]
    pack: Pack
    persona: Persona
    banished: Banished = Field(default_factory=Banished)
    rules: Rules = Field(default_factory=Rules)
    pop_culture: PopCulture = Field(default_factory=PopCulture)
    formats: list[Format] = Field(default_factory=list)
    samples: list[Sample] = Field(default_factory=list)
    bios: list[Bio] = Field(default_factory=list)
